webapp/channels/src/workers/sw.js

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+// Update this every time code is changed.
+// This is needed to delete old cache entries.
+const cacheName = 'v1';
+const ROOT_HTML_REQUEST = new Request('root.html');
+
+const putInCache = async (request, response) => {
+    const cache = await caches.open(cacheName);
+    if (request.destination === 'document') {
+        await cache.put(ROOT_HTML_REQUEST, response);
+    } else {
+        await cache.put(request, response);
+    }
+};
+
+const handleCacheRequest = async (request) => {
+    // Using the stale-while-revalidate approach for cached requests to keep them in sync.
+    let responseFromCache;
+    if (request.destination === 'document') {
+        // If it's an HTML document, then we check for a separate key.
+        responseFromCache = await caches.match(ROOT_HTML_REQUEST);
+    } else {
+        responseFromCache = await caches.match(request);
+    }
+    if (responseFromCache) {
+        // Non-blocking cache insertion.
+        fetchFromNetwork(request);
+        return responseFromCache;
+    }
+
+    // Wait till we get a response.
+    const responseFromNetwork = await fetchFromNetwork(request);
+    return responseFromNetwork;
+};
+
+const fetchFromNetwork = async (request) => {
+    try {
+        const responseFromNetwork = await fetch(request.clone());
+        putInCache(request, responseFromNetwork.clone());
```

Comment:
Do we need to do anything to avoid caching error responses here? I can't remember if `fetch` throws an error when a server returns something like a 404 or if it only throws those for network errors